### PR TITLE
insert_deduplication_token setting for INSERT statement

### LINKED ISSUE #7461: Store user defined insert_id in zookeper
**Use case**

I want to gurantee exactly once delivery to clickhouse without requiring separate persistence for replicated merge tree families when i use upload from some message broker (kafka-like).
Most message brokers has stream api. We read buffer of data, push it to clickhouse and commit offsets when it done. So in case of failures such client code may experience double read. If worker push data to clikchouse it may lead to double writes aswell. Clickhouse already has well defined deduplication mechanism, clickhouse will dedublicate data if you push exact same batch multiple time. So in our worker (assuming we write in single thread to CH) we must know last request offsets. To achive these i have to use separate storage and do 2 writes into it (before insert with right offset, and after ack from CH to confirm left offset). 

User flow looks something like this:
1. Read data from broker (let say 1000 record, from N to M offset)
2. Store right ofset to M
3. Start upload data to clickhouse
4. Die in agony
5. Restore worker, look to offset store, and see that last time there was N to M offset sent
6. Read from broker exact same offsets
7. Push it again to clickhouse
8. Store M offset to left.

**Describe the solution you'd like**

Main idea to eliminate need of separate storage. Right now clickhouse already utilize persistence via ZK, so we could use it for these need. If user will provide some meaningfull (for him) string with each write request we can consistent store it to ZK. Then on client side we may look at it and make a decision of what offset may be skipped.

Ideal user flow looks something like this:
1. Read data from broker from N to M.
2. Start upload to clickhouse with insert id N_M
3. Die with happy smile
4. Restore worker. Look to clickhosue to latest insert id (N_M recieved)
5. Skip offset till M 
6. Go to 1

**Describe alternatives you've considered**
We could continue to use separate offset store.

**Additional context**
It may be also usefull for internal use of kafka storage (wich also have same proble).

### PR DESCRIPTION
**Pull request by devcrafter**
The setting allows a user to provide own deduplication semantic in MergeTree/ReplicatedMergeTree
If provided, it's used instead of data digest to generate block ID
So, for example, by providing a unique value for the setting in each INSERT statement,
user can avoid the same inserted data being deduplicated

Inserting data within the same INSERT statement are split into blocks
according to the *insert_block_size* settings
(max_insert_block_size, min_insert_block_size_rows, min_insert_block_size_bytes).
Each block with the same INSERT statement will get an ordinal number.
The ordinal number is added to insert_deduplication_token to get block dedup token
i.e. <token>_0, <token>_1, ... Deduplication is done per block
So, to guarantee deduplication for two same INSERT queries,
dedup token and number of blocks have to be the same

This closes: #7461

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The setting allows a user to provide own deduplication semantic in MergeTree/ReplicatedMergeTree If provided, it's used instead of data digest to generate block ID. So, for example, by providing a unique value for the setting in each INSERT statement, the user can avoid the same inserted data being deduplicated. This closes: #7461

Detailed description / Documentation draft:

TODO

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.
> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.
> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
@devcrafter Could you please allow edits to your branch?

File: src/Storages/MergeTree/IMergeTreeDataPart.h [hunk 1]
```diff
@@ -170,7 +170,8 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
     bool isEmpty() const { return rows_count == 0; }
 
     /// Compute part block id for zero level part. Otherwise throws an exception.
-    String getZeroLevelPartBlockID() const;
+    /// If token is not empty, block id is calculated based on it instead of block data
+    String getZeroLevelPartBlockID(const String & token = String()) const;
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/IMergeTreeDataPart.h:174]: Default value is unnecessary.

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp [hunk 2]
```diff
@@ -160,8 +160,16 @@ void ReplicatedMergeTreeSink::consume(Chunk chunk)
...
             /// We add the hash from the data and partition identifier to deduplication ID.
             /// That is, do not insert the same data to the same partition twice.
-            block_id = part->getZeroLevelPartBlockID();
 
+            String block_dedup_token = context->getSettingsRef().insert_deduplication_token;
+            if (!block_dedup_token.empty())
+            {
+                /// multiple blocks can be inserted within the same insert query
+                /// an ordinal number is added to dedup token to generate a distinctive block id for each block
+                block_dedup_token += fmt::format("_{}", chunk_dedup_seqnum++);
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:169]: Let's split into two statements.

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp [hunk 3]
```diff
@@ -160,8 +160,16 @@ void ReplicatedMergeTreeSink::consume(Chunk chunk)
...
+                LOG_TRACE(log, "Deduplication token: {}", block_dedup_token);
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:170]: Let's remove.

File: src/Storages/MergeTree/IMergeTreeDataPart.h (hunk 1 above)
Comment (devcrafter) [reply src/Storages/MergeTree/IMergeTreeDataPart.h:174]: Yes, but after deduplication token will be added to non-replicated MergeTree  as well

**Comment by devcrafter** [comment]
> @devcrafter Could you please allow edits to your branch?

Done

**Comment by devcrafter** [comment]
@alexey-milovidov I still want to complete the PR with dedup token for non-replicated MergeTree. Don't you mind assigning it to me?

**Comment by alexey-milovidov** [comment]
Yes, sure. I expect both Replicated and non-Replicated can be implemented in this PR.

**Comment by alesapin** [comment]
@mergifyio update

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
GitHub is notoriously unstable in the last day, so CI is delayed:
https://news.ycombinator.com/item?id=30109187

**Pull request merged**